File: HW5 - C++/BoardArray2D.cpp
#include <iostream>
#include "AbstractBoard.h"
#include "BoardArray2D.h"
using namespace std ;

namespace GTU_CSE241_2019_2020{
	

	BoardArray2D::BoardArray2D() : AbstractBoard() , board(nullptr){
		// empty
	}

	BoardArray2D::BoardArray2D( const BoardArray2D &other ) : AbstractBoard(){
		
		setSize( other.row, other.column ) ;

		for (int i = 0; i < row; ++i)
			for (int j = 0; j < column; ++j)
				this->board[i][j] = other.board[i][j] ;

	}

	BoardArray2D& BoardArray2D::operator =( const BoardArray2D  &rightSide ){
		
		if( *this == rightSide )
			return *this ;
		setSize( rightSide.row, rightSide.column ) ;

		for (int i = 0; i < row; ++i)
			for (int j = 0; j < column; ++j)
				this->board[i][j] = rightSide.board[i][j] ;

		return *this ;

	}
	void BoardArray2D::setSize( int new_row, int new_column ){

		if( new_row < 3 || new_column < 3 ){ // set as default

			for (int i = 0; i < row; ++i)
				delete [] board[i] ;
			delete [] board ;

			row = 3 ;
			column = 3 ;

			board = new int * [row] ;
			for (int i = 0; i < row; ++i)
				board[i] = new int [column] ;

		}
		else if( new_row != row || new_column != column ){

			for (int i = 0; i < row; ++i)
				delete [] board[i] ;
			delete [] board ;

			row = new_row ;
			column = new_column ;

			board = new int * [row] ;
			for (int i = 0; i < row; ++i)
				board[i] = new int [column] ;
		}
		
		reset() ;
	}

	
	
	int& BoardArray2D::operator () ( int row_cordinate, int column_cordinate ){

		return board[ row_cordinate ][ column_cordinate ] ;
	}

	const int& BoardArray2D::operator () ( int row_cordinate, int column_cordinate )const{

		return board[ row_cordinate ][ column_cordinate ] ;

	}
	
	BoardArray2D::~BoardArray2D(){
		for (int i = 0; i < row; ++i)
				delete [] board[i] ;
			delete [] board ;
	}

}

File: HW7 - JAVA/src/HW7/BoardArray2D.java
package HW7;

/**
 * @author murat
 * @since 15.01.2020
 *
 */

/**
 * A class which to represent NPuzzle problem.It keeps data as two dimensional integer array.
 * @see AbstractBoard
 */
public class BoardArray2D extends AbstractBoard {
	/** 
	 * Keep data in two dimensional integer array.
	 */
	private int [][] board ;
	
	/** 
	 * Initialize fields
	 * <p> Calls AbstractBoard() as a  super class constructor
	 * @see AbstractBoard#AbstractBoard()
	 */
	public BoardArray2D(){
		super() ;	
	}
	

	@Override
	public void setSize( int newRow, int newColumn ) {
		
		if( board != null && !( newRow == row && newColumn == column) ) {	

			board = null ;
			System.gc();
		}
		
		if( newRow < 3 || newColumn < 3 ){ // set as default
			
			row = 3 ;
			column = 3 ;
			board = new int [row][column] ;

		}
		else if( newRow != row || newColumn != column ){

			row = newRow ;
			column = newColumn ;
			board = new int [row][column] ;
		}
		
		super.reset() ;
	}

	@Override
	public int cell(int rowIndex, int columnIndex) {
		
		return board[rowIndex][columnIndex] ;
	}

	@Override
	public void setCell(int rowIndex, int columnIndex, int setValue) {
		
		board[rowIndex][columnIndex] = setValue ;
	}
	
	
}
